src/gr_mat/randsimilar.c

```diff
@@ -0,0 +1,76 @@
+/*
+    Copyright (C) 2025 Ricardo Buring
+
+    This file is part of FLINT.
+
+    FLINT is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.  See <https://www.gnu.org/licenses/>.
+*/
+
+#include "gr.h"
+#include "gr_vec.h"
+#include "gr_mat.h"
+
+int
+gr_mat_randsimilar(gr_mat_t mat, flint_rand_t state, slong opcount, gr_ctx_t ctx)
+{
+    slong c, i, j;
+    slong m = mat->r;
+    slong n = mat->c;
+    slong sz = ctx->sizeof_elem;
+    gr_mat_t W1, W2;
+    gr_ptr x;
+    int status = GR_SUCCESS;
+
+    if (n != m)
+        return GR_DOMAIN;
+
+    if (n <= 1)
+        return GR_SUCCESS;
+
+    GR_TMP_INIT(x, ctx);
+
+    /* Conjugate by random elementary operations. */
+    for (c = 0; c < opcount; c++)
+    {
+        switch (n_randint(state, 3)) {
+            case 0: {
+                /* Swap row i and row j, swap column i and column j. */
+                if ((i = n_randint(state, n)) == (j = n_randint(state, n)))
+                    continue;
+                status |= gr_mat_swap_rows(mat, NULL, i, j, ctx);
+                status |= gr_mat_swap_cols(mat, NULL, i, j, ctx);
+            } break;
+            case 1: {
+                /* For a random invertible x, scale row i by x, scale column i by the inverse of x. */
+                i = n_randint(state, n);
+                do {
+                    status |= gr_randtest_small(x, state, ctx);
+                } while (gr_is_invertible(x, ctx) != T_TRUE);
```

Comment:
I guess this loop needs an escape since it will hang if the ring doesn't implement ``is_invertible`` or if it happens to be a nontrivial implementation of the zero ring which doesn't recognize 0 as invertible or something like that.

Reply:
I guess it would be useful to have ``gr_randtest_invertible`` which does something like ``gr_generic_randtest_not_zero`` or the ad hoc code I put in ``gr_test_canonical_associate``: call ``gr_randtest`` and check invertibility up to 5 times, and as a fallback randomly return 1 or -1 (presumed to be invertible in any ring). Would you like to add such a helper function as part of this PR?

Reply:
Yes, I agree a `gr_randtest_invertible` would be useful generally.

I'll have a look at those options you mentioned and give it a shot in this PR.